### ISSUE TITLE
ESLint: 'className' is missing in props validation(react/prop-types)

### ISSUE THREAD

**Issue by 1August** [shadcn-ui/ui#1013]
When using ESLint you can get error "ESLint: 'className' is missing in props validation(react/prop-types)". Just need to add interface that extends React.HTMLAttributes.
Was:

`const Card = React.forwardRef<HTMLDivElement, React.HTMLAttributes<HTMLDivElement>>
(({ className /* warning is here */, ...props }, ref) => (...))`

Should be:
**interface CardProps extends React.HTMLAttributes<HTMLDivElement> {}**

`const Card = React.forwardRef<HTMLDivElement, **CardProps**>
(({ className, ...props }, ref) => (...))`

**Comment by seongsoo96** [comment]
every shadcn-ui components got error of react.prop-types. so i add
    "react/prop-types": [2, { "ignore": "className" }],
to my rule of .eslintrc.
After that we don't need to add extra interface every shadcn-ui components.

if there is any better way, mention me please!

**Comment by bbc-buers** [comment]
> [...] so i add "react/prop-types": [2, { "ignore": "className" }], to my rule of .eslintrc [...]

Just as a side note, the value of `ignore` is supposed to be an array, e.g. `"ignore": ["className"]`

**Comment by GOWxx** [comment]
Add this content to the eslint configuration file

```cjs
module.exports = {
// ...
  overrides: [
    // React
    {
      files: ["**/*.{js,jsx,ts,tsx}"],
      // ...
      rules: {
        "react/prop-types": [2, {
          "ignore": ["className", "variant", "size"]
        }]
      }
    },
  // ...
  ],
};
```

**Comment by fernandocanizo** [comment]
> every shadcn-ui components got error of react.prop-types. so i add "react/prop-types": [2, { "ignore": "className" }], to my rule of .eslintrc. After that we don't need to add extra interface every shadcn-ui components.
> 
> if there is any better way, mention me please!

Ignoring a rule to shut up a valid Typescript error is not a fix, if we are going to start ignoring whatever makes noise, then just let's take out Eslint and Typescript from our projects.

The proper way to fix it is to define the `className` or any other parameter in the type definitions where it's required.

Something along this pattern:
```ts
export default function DatePickerWithRange({
  className,
}: React.HTMLAttributes<HTMLDivElement> & {
  className?: string
}) {
// ...
```

It would be better if this could be addressed at the library level, as I've seen this error in many components.

**Comment by shadcn** [comment]
This issue has been automatically closed because it received no activity for a while. If you think it was closed by accident, please leave a comment. Thank you.

**Comment by driera** [comment]
This is still happening. Importing `HTMLAttributes` type from react and using it solves the issue:

```typescript
import * as React from "react";
import { cn } from "@/lib/utils";
import type { HTMLAttributes } from "react";

const Card = React.forwardRef<
  HTMLDivElement, HTMLAttributes<HTMLDivElement>
>(({ className, ...props }, ref) => (
// ...
```

**Comment by benrbowers** [comment]
My suggestion is to turn off "react/prop-types", but ONLY for shadcn/ui

Add to eslint config:
```javascript
module.exports = {
  overrides: [
    ...,
    // shadcn/ui
    {
      files: ['app/shadcn/ui/**/*.{ts,tsx}'],
      rules: {
	"react/prop-types": 'off'
      }
    }
  ]
};
```

**Comment by advaitju** [comment]
Ignoring the rule in eslint is a terrible idea because you're essentially disabling your linting by disabling this rule: `"react/prop-types": 'off'`

You may well need it in another part of your app.

This is issue is caused because somewhere in the typing `className` isn't being found.

A better solution is to extend the prop type/interface with the missing types.

```tsx
// Option 1 (seems better): Get exact attributes types
const ToastViewport = React.forwardRef<
  React.ElementRef<typeof ToastPrimitives.Viewport>,
  React.ComponentPropsWithoutRef<typeof ToastPrimitives.Viewport> &
    React.HTMLAttributes<typeof ToastPrimitives.Viewport>
```

```tsx
// Option 2: Use default JSX attributes types
const ToastViewport = React.forwardRef<
  React.ElementRef<typeof ToastPrimitives.Viewport>,
  React.ComponentPropsWithoutRef<typeof ToastPrimitives.Viewport> &
    React.JSX.IntrinsicAttributes
```

**Comment by rbondoc96** [comment]
I encountered this issue recently after doing some refactoring and came across this issue. Posting my approach, in case someone finds it useful.

#### Versions (from `package-lock.json`)
```
* @types/react: 18.3.12
* eslint: 9.14.0
* eslint-plugin-react: 7.37.2
* typescript: 5.6.3
```

#### TLDR
If you really want to use the `import * as React from 'react'` pattern and you don't want to mess with the `react/prop-types` rule in your ESLint config, the path I decided to go with was:

* For HTML elements, use `React.JSX.IntrinsicElements['tag']` over `React.ComponentProps<'tag'>`
    * When used with `forwardRef`, `ComponentProps` is functionally the same as using `ComponentPropsWithoutRef` 
* For Radix components, use the prop types for each element whenever possible.

> [!TIP]
> Some things that go on under the hood in `@types/react`:
>
> * `ComponentPropsWithoutRef<T>` is a type alias for `PropsWithoutRef<ComponentProps<T>>`
> * `ComponentProps<T>` is a type alias that uses `JSX.IntrinsicElements[T]`, if `T extends keyof JSX.IntrinsicElements`
> * `forwardRef<T, P>` uses `PropsWithoutRef<P>`, so the type of `ref` on `P` isn't actually used to type the `ref` in the render callback

#### Example

As an example, with the `Sheet` component:
```tsx
import * as SheetPrimitive from '@radix-ui/react-dialog';
import * as React from 'react';

// With a Radix component, its props type can be used (hopefully it's available)
export const SheetDescription = React.forwardRef<
    React.ElementRef<typeof SheetPrimitive.Description>,
    SheetPrimitive.DialogDescriptionProps
>(({ children, className, ...props }, ref) => {
    return (
        <DialogPrimitive.Description className={cn('text-sm text-muted-foreground', className)} ref={ref} {...props}>
            {children}
        </DialogPrimitive.Description>
    );
});
SheetDescription.displayName = 'SheetDescription';

// For a component that uses HTML elements, `JSX.IntrinsicElements` can be used
export const SheetHeader = React.forwardRef<HTMLDivElement, React.JSX.IntrinsicElements['div']>(
    ({ children, className, ...props }, ref) => {
        return (
            <div className={cn('flex flex-col gap-2 text-center sm:text-left', className)} ref={ref} {...props}>
                {children}
            </div>
        );
    },
);
SheetHeader.displayName = 'SheetHeader';
```

How I got to my current approach is explained below. Why this is an issue is beyond my current understanding, but these are just things I noticed.

---

<details>
<summary>(Click to expand) I wanted to simplify the import structure in my code, seeing where I could safely use namespace imports instead of explicitly importing everything I needed by name.</summary>

<br>

The ESLint error initially didn't occur when I brought in <code>ComponentPropsWithoutRef</code> as a named import. But once I transitioned to namespace imports... 🫠

Initially, I was using `ComponentPropsWithoutRef` as a named import:
```tsx
// Depending on the file, this import can get pretty long.
import {type ComponentPropsWithoutRef, type ElementRef, forwardRef} from 'react';

export const Foo = forwardRef<HTMLDivElement, ComponentPropsWithoutRef<'div'>(
    // No ESLint error here for the `className` prop.
    ({ className, ...props }, ref) => {
        //
    },
);
```

But once I switched over to using a namespace import...
```tsx
// Also occurs if using `import React from 'react'`
import * as React from 'react';

export const Foo = React.forwardRef<HTMLDivElement, React.ComponentPropsWithoutRef<'div'>(
    // Oh no, ESLint error :(
    ({ className, ...props }, ref) => {
        //
    },
);
```
</details>
